File: linkedList/doublyLL.cpp
#include<bits/stdc++.h>
using namespace std;

//considering 0 based indexing

class Node{
public:
    int val;
    Node* next;
    Node* prev;

    Node(int val){
        this->val = val;
        next = prev = NULL;

    }
};

void insert(Node* &head,int val, int pos){
    Node* nodeToBeInserted = new Node(val);
    if(head == NULL || pos <0 ){
        cout << "invalid pos or no linked list";
        return;
    }

    if(pos == 0){
        nodeToBeInserted->next = head;
        if (head) head->prev = nodeToBeInserted;
        head = nodeToBeInserted;
        return;
    }

    
    Node* temp = head;
    for(int i = 0; i<pos-1 && temp != NULL ; i++){
        temp=temp->next;
    }
    if(temp==NULL){
        cout << "Not possible to insert";
        return;
    }
    Node* nextNode = temp->next;
    temp->next = nodeToBeInserted;
    nodeToBeInserted->next = nextNode;
    nodeToBeInserted->prev = temp;
    nextNode->prev = nodeToBeInserted;
}


void deleteNode(Node* &head, int pos){
     if(head == NULL || pos < 0 ){
        cout << "invalid pos or no linked list";
        return;
    }

    if(pos == 0){
        Node* toDelete = head;
        head = head->next;
        if (head) head->prev = NULL;
        delete toDelete;
        return;
    }


    Node* temp = head;
    for(int i = 0; i<pos-1 && temp!=NULL ; i++){
        temp = temp->next;
    }
    if(temp==NULL || temp->next==NULL){
        cout << "Not possible to delete";
        return;
    }
    Node* toDelete = temp->next;
    temp->next = toDelete->next;
    toDelete->prev = temp;
    delete toDelete;

}



void printLL(Node* node){
    while(node){
        if(node->next == NULL){
             cout << node->val;
        }
        else cout << node->val << " <-> ";
        node=node->next;
    }
    cout << endl;
}


int main(){
    Node* head = new Node(10);
    Node* x = new Node(20);
    Node* y = new Node(30);
    head->next = x;
    x->prev = head;
    x->next = y;
    y->prev = x;

    printLL(head);

    insert(head,25,0);//0 based

    printLL(head);

    deleteNode(head,0);

    printLL(head);



    return 0;
}

File: linkedList/doublyLL.java
class DoublyLL {

    static class Node {
        int val;
        Node next;
        Node prev;

        Node(int val) {
            this.val = val;
            next = null;
            prev = null;
        }
    }

    // Insert at position (1-based index)
    public static Node insert(Node head, int val, int pos) {
        Node newNode = new Node(val);

        if (pos <= 0) return head;

        if (pos == 1) {
            newNode.next = head;
            if (head != null) head.prev = newNode;
            return newNode;
        }

        Node curr = head;
        for (int i = 1; i < pos - 1 && curr != null; i++) {
            curr = curr.next;
        }

        if (curr == null) return head; // position out of bounds

        newNode.next = curr.next;
        newNode.prev = curr;
        if (curr.next != null) curr.next.prev = newNode;
        curr.next = newNode;

        return head;
    }

    // Delete node at position (1-based index)
    public static Node deleteNode(Node head, int pos) {
        if (head == null || pos <= 0) return head;

        if (pos == 1) {
            Node nextNode = head.next;
            if (nextNode != null) nextNode.prev = null;
            return nextNode;
        }

        Node curr = head;
        for (int i = 1; i < pos && curr != null; i++) {
            curr = curr.next;
        }

        if (curr == null) return head; // position out of bounds

        if (curr.prev != null) curr.prev.next = curr.next;
        if (curr.next != null) curr.next.prev = curr.prev;

        return head;
    }

    // Print the linked list
    public static void printLL(Node head) {
        while (head != null) {
            System.out.print(head.val + " <-> ");
            head = head.next;
        }
        System.out.println("null");
    }

    public static void main(String[] args) {
        Node head = null;

        head = insert(head, 10, 1);
        head = insert(head, 20, 2);
        head = insert(head, 30, 3);
        printLL(head);

        head = insert(head, 15, 2); // insert 15 at position 2

        System.out.print("After insertions: ");
        printLL(head);

        head = deleteNode(head, 3); // delete node at position 3

        System.out.print("After deletion at pos 3: ");
        printLL(head);
    }
}
